File: frontend/src/components/benchmark/content/eigenvector_centrality/implementations/eigenvector_centrality.java
import java.util.*;

public class EigenvectorCentrality {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public Map<Integer, Double> eigenvectorCentrality(int maxIter, double tol) {
        Map<Integer, Double> centrality = new HashMap<>();
        int n = nodes.size();

        for (int node : nodes) {
            centrality.put(node, 1.0 / n);
        }

        for (int iter = 0; iter < maxIter; iter++) {
            Map<Integer, Double> newCentrality = new HashMap<>();
            double norm = 0.0;

            for (int node : nodes) {
                double sum = 0.0;
                for (int neighbor : adjacency.get(node)) {
                    sum += centrality.get(neighbor);
                }
                newCentrality.put(node, sum);
                norm += sum * sum;
            }

            norm = Math.sqrt(norm);
            double diff = 0.0;

            for (int node : nodes) {
                newCentrality.put(node, newCentrality.get(node) / norm);
                diff += Math.abs(newCentrality.get(node) - centrality.get(node));
            }

            centrality = newCentrality;

            if (diff < tol) {
                break;
            }
        }

        return centrality;
    }

    public List<Map.Entry<Integer, Double>> topKCentralNodes(int k, int maxIter, double tol) {
        Map<Integer, Double> centrality = eigenvectorCentrality(maxIter, tol);
        List<Map.Entry<Integer, Double>> result = new ArrayList<>(centrality.entrySet());

        result.sort((a, b) -> b.getValue().compareTo(a.getValue()));

        return result.subList(0, Math.min(k, result.size()));
    }

    public Map<Integer, Double> normalizedEigenvectorCentrality(int maxIter, double tol) {
        Map<Integer, Double> centrality = eigenvectorCentrality(maxIter, tol);
        double maxVal = 0.0;

        for (double value : centrality.values()) {
            if (value > maxVal) {
                maxVal = value;
            }
        }

        if (maxVal > 0) {
            for (int node : nodes) {
                centrality.put(node, centrality.get(node) / maxVal);
            }
        }

        return centrality;
    }
}

File: frontend/src/components/benchmark/content/eigenvector_centrality/implementations/eigenvector_centrality.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <cmath>

using namespace std;

class EigenvectorCentrality {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;

public:
    void add_edge(int u, int v) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
    }

    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (const auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }

    unordered_map<int, double> eigenvector_centrality(int max_iter = 100, double tol = 1e-6) {
        unordered_map<int, double> centrality;
        int n = nodes.size();

        for (int node : nodes) {
            centrality[node] = 1.0 / n;
        }

        for (int iter = 0; iter < max_iter; iter++) {
            unordered_map<int, double> new_centrality;
            double norm = 0.0;

            for (int node : nodes) {
                double sum = 0.0;
                for (int neighbor : adjacency[node]) {
                    sum += centrality[neighbor];
                }
                new_centrality[node] = sum;
                norm += sum * sum;
            }

            norm = sqrt(norm);
            double diff = 0.0;

            for (int node : nodes) {
                new_centrality[node] /= norm;
                diff += abs(new_centrality[node] - centrality[node]);
            }

            centrality = new_centrality;

            if (diff < tol) {
                break;
            }
        }

        return centrality;
    }

    vector<pair<int, double>> top_k_central_nodes(int k = 10, int max_iter = 100, double tol = 1e-6) {
        auto centrality = eigenvector_centrality(max_iter, tol);
        vector<pair<int, double>> result(centrality.begin(), centrality.end());

        sort(result.begin(), result.end(),
             [](const auto& a, const auto& b) {
                 return a.second > b.second;
             });

        if (k > (int)result.size()) {
            k = result.size();
        }

        return vector<pair<int, double>>(result.begin(), result.begin() + k);
    }

    unordered_map<int, double> normalized_eigenvector_centrality(int max_iter = 100, double tol = 1e-6) {
        auto centrality = eigenvector_centrality(max_iter, tol);
        double max_val = 0.0;

        for (const auto& pair : centrality) {
            if (pair.second > max_val) {
                max_val = pair.second;
            }
        }

        if (max_val > 0) {
            for (auto& pair : centrality) {
                pair.second /= max_val;
            }
        }

        return centrality;
    }
};